crates/storage-query-datafusion/src/table_util.rs

```diff
@@ -55,7 +56,45 @@ pub(crate) trait Builder {
 
     fn empty(&self) -> bool;
 
-    fn finish(self) -> datafusion::common::Result<RecordBatch>;
+    fn finish(&mut self) -> datafusion::common::Result<RecordBatch>;
 
     fn finish_and_new(&mut self) -> datafusion::common::Result<RecordBatch>;
 }
+
+pub(crate) struct BatchSender<B: Builder> {
+    builder: B,
+    tx: Sender<datafusion::error::Result<RecordBatch>>,
+}
+
+impl<B: Builder> BatchSender<B> {
+    pub(crate) fn new(
+        projection: SchemaRef,
+        tx: Sender<datafusion::error::Result<RecordBatch>>,
+    ) -> Self {
+        let builder = B::new(projection);
+        Self { builder, tx }
+    }
+
+    pub(crate) fn send(
+        &mut self,
+    ) -> Result<(), tokio::sync::mpsc::error::SendError<datafusion::error::Result<RecordBatch>>>
+    {
+        self.tx.blocking_send(self.builder.finish_and_new())
+    }
+
+    pub(crate) fn builder_mut(&mut self) -> &mut B {
+        &mut self.builder
+    }
+
+    pub(crate) fn num_rows(&self) -> usize {
+        self.builder.num_rows()
+    }
+}
+
+impl<B: Builder> Drop for BatchSender<B> {
+    fn drop(&mut self) {
+        if !self.builder.empty() {
+            let _ = self.tx.blocking_send(self.builder.finish());
```

Comment:
Perhaps wrap the builder in ManuallyDrop<> instead of changing finish() to accept `&mut self`?

Reply:
huh, thats how i did it originally but i thought we'd want to avoid the unsafe. ok, will do